### ISSUE TITLE
`Scope` ability and local mutable references

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#2340]
Allow local mutation in pure code, with an `ST`-like interface to prevent it from being observed outside some scope.

Needs some more design, @dolio we should jam on this a bit, but a while ago I started sketching this, and it comes out quite nice using abilities -

https://github.com/unisonweb/unison/issues/798#issuecomment-702487151

We’d need `Scope.Ref` (aka `STRef`) as well.

This is useful for the distributed API, for local pure testing, but it’s also just generally useful for other stuff.

**Comment by pchiusano** [comment]
Latest design: 

```Haskell
builtin unique ability Scope t 
builtin type Ref g a = ...
Ref.get : Ref g a ->{g} a
Ref.set : a -> Ref g a ->{g} ()
Scope.ref : a -> Ref {Scope t} a
IO.ref : a -> Ref {IO} a
```

**Comment by pchiusano** [comment]
Fixed by #2351
